src/Blocks/BlockEndPortalFrame.h

```diff
@@ -0,0 +1,239 @@
+			Vector3i( 1, 0, -1),  // 3, East,  left block is NorthEast / XP ZM
+		};
+
+
+		int Turns = 0;
+		int Proper = 1;  // The first portal frame was already checked.
+		Vector3i Former = a_FirstFrame;
+		Vector3i Corner;  // The first and third corners.
+		int Width[4] = {1, 1, 1, 1};
+
+		while (Turns <= 4)
+		{
+			// Since we start on the first wall, get the corner, then cover each wall, we're effectively scanning 5 walls.
+			if (Proper > (MAX_PORTAL_WIDTH * 5))
+			{
+				return 0;
+			}
+
+			int CornerValid = IsValidFrame(Former + LeftForward[(a_Direction + Turns) % 4], (a_Direction + Turns + 1) % 4, a_ChunkInterface);
+			if (CornerValid == E_VALID_PORTAL_FRAME)
+			{
+				Former = Former + LeftForward[(a_Direction + Turns) % 4];
+				if ((a_Direction + Turns) % 4 == 3)
+				{
+					// Remember the NW (XM, ZM) Corner
+					// Relative to the previous frame, the portal should appear to the right of the new portal frame.
+					Corner = Former - Left[(a_Direction + Turns) % 4];
+				}
+				Turns++;
+				Proper++;
+				continue;
+			}
+			else if (CornerValid == E_INVALID_PORTAL_FRAME)
+			{
+				// The Portal Frame existed, but was either missing the eye, or was facing the wrong direction.
+				return 0;
+			}
+			else
+			{
+
+				// Corner is not a portal block, try edge.
+				int EdgeValid = IsValidFrame(Former + Left[(a_Direction + Turns) % 4], (a_Direction + Turns) % 4, a_ChunkInterface);
+				if (EdgeValid == E_VALID_PORTAL_FRAME)
+				{
+					if (Turns >= 1)
+					{
+						Width[(a_Direction + Turns) % 4]++;
+
+						if (Width[(a_Direction + Turns) % 4] > MAX_PORTAL_WIDTH)
+						{
+							// Don't build a portal that is too long.
+							return 0;
+						}
+					}
+					Former = Former + Left[(a_Direction + Turns) % 4];
+					Proper++;
+					continue;
+				}
+				else if (EdgeValid == E_INVALID_PORTAL_FRAME)
+				{
+					return 0;
+				}
+				else
+				{
+					// Neither the edge nor the corner was a portal block.
+					return 0;
+				}
+			}
+		}
+
+		if ((Width[0] != Width[2]) || (Width[1] != Width[3]))
+		{
+			// Mismatched Portal Dimensions.
+			return 0;
+		}
+		if ((Width[0] < MIN_PORTAL_WIDTH) || (Width[1] < MIN_PORTAL_WIDTH))
+		{
+			// Portal too small.
+			return 0;
+		}
+
+		// LOG("NW corner (low corner) %d %d %d", Corner.x, Corner.y, Corner.z);
+		// LOG("%d by %d", Width[0], Width[1]);
+		for (int i = 0; i < Width[0]; i++)
+		{
+			for (int j = 0; j < Width[1]; j++)
+			{
+				a_ChunkInterface.SetBlock(Corner.x + i, Corner.y, Corner.z + j, E_BLOCK_END_PORTAL, 0);
+				// TODO: Create block entity so portal doesn't become invisible on relog.
+			}
+		}
+		return 1;
+	}
+
+
+
+
+
+	/** Check a specific block to see if it is a valid portal frame. */
+	int IsValidFrame(Vector3i a_Pos, NIBBLETYPE a_ShouldFace, cChunkInterface & a_ChunkInterface)
+	{
+		BLOCKTYPE BlockType;
+		NIBBLETYPE BlockMeta;
+		if (! a_ChunkInterface.GetBlockTypeMeta(a_Pos, BlockType, BlockMeta))
```

Comment:
Style: spaces before the ! are not good.

Reply:
There is no space before this `!`.

Since it's an operator, I have a space after it.

Reply:
Sorry, a slip of the fingers. I meant after. Even though `!` is an operator, it is an exception to the general rule and has no spaces before or after.